Warn when replacing placeholders

graemeboyd opened the pull request:
This PR addresses an issue we've faced with a large Turnip codebase using placeholders with generic names like `:object` in multiple helpers across multiple files. This led to the placeholders being easily overridden and producing obscure and hard to track down bugs in our test suite.

The main change is to issue a warning that a a placeholder is being replaced if there's already a non-default placeholder. The warning includes the file and line using `caller_locations` which is supported in Ruby 2.2 onwards.

With just this change the Turnip test suite now issues warnings as it runs as it relies on replacing placeholders being silent. To address this I've added a `.clear` convenience method which removes all the added placeholders reverting them to the original default.

hanachin commented on lib/turnip/placeholder.rb:
```diff
@@ -19,6 +20,10 @@ def find(name)
         placeholders[name] or default
       end
 
+      def clear
```
IMHO adding a new public API to eliminate warnings doesn't make sense to me.
How about define the `clear` method in a Refinement and using it in the test instead?

@gongo What do you think?

```rb
# spec/support/clear_place_holders
module ClearPlaceHolders
  refine(Turnip::Placeholder) do
    def clear
      placeholders.clear
    end
  end
end

# spec/placeholder_spec.rb
using ClearPlaceHolders

describe Turnip::Placeholder do
  after do
    described_class.clear
  end
  # ...
end
```

gongo replied:
@hanachin Thanks! I think it's good
@graemeboyd I don't think it's good to make it Public API if `Placeholder.clear` just code for testing.

graemeboyd commented on lib/turnip/placeholder.rb:
```diff
@@ -4,6 +4,7 @@ class Match < Struct.new(:regexp, :block);end
 
     class << self
       def add(name, &block)
+        warn "Placeholder :#{name} was replaced at #{caller_locations.first}." if placeholders.key?(name)
```
Just realised this doesn't give the correct `calller_location` when called with the dsl.

graemeboyd commented:
Hi @gongo and @hanachin, sorry it's taken me a while to continue this. I hadn't realised clearing the placeholder hash by using send was already used elsewhere in the test suite. I've changed the PR to remove introducing a new public method and I've squashed the commits down.

gongo commented:
@graemeboyd @hanachin Thanks for good change and comments!
